[1-3단계 - 방탈출 예약 관리] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
https://prolog.techcourse.co.kr/studylogs/4153

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.


## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
안녕하세요 샐리~ 이번 미션 리뷰이인 레오입니다.
리뷰 잘부탁드려요!

스프링을 활용한 프로젝트 경험이 2번정도 있지만, 그냥 인강에서 보고 습득한 기술들을 소위 말해 깔짝깔짝 사용만 해봤다는 느낌이 강해요! 그래서 이번 미션에서는 과거에 작성해보았던 코드들도 다시 공식 문서를 통해 사용법을 익혀가며 문서를 통해 기술을 습득하는 방법을 배워가고자 합니다.

리뷰는 전체적으로 봐주시면 감사하겠습니다!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

jjunh0 committed:
docs: 1단계 요구사항 작성

jjunh0 committed:
feat: 웰컴 페이지 추가

jjunh0 committed:
feat: 어드민 페이지 추가

kjyyjk committed:
docs: 2단계 요구사항 작성

kjyyjk committed:
feat: 예약 관리 페이지 응답

kjyyjk committed:
feat: 예약 목록 조회 API 구현

- 사용자가 원하는 데이터가 `Reservation`과 일치하지 않을 수 있기 때문에 Dto를 반환한다.

jjunh0 committed:
docs: 3단계 요구사항 작성

jjunh0 committed:
feat: 예약 추가 API 구현

jjunh0 committed:
chore: 시간 형식 변경

kjyyjk committed:
feat: 예약 취소 API 구현

kjyyjk committed:
style: 메서드명 수정

- Controller 클래스 명에 도메인이 명시되어있으므로
메서드의 도메인 명시 부분을 제거한다

kjyyjk committed:
style: 개행 삭제 및 추가

kjyyjk commented on src/main/java/roomescape/ReservationController.java:
```diff
@@ -0,0 +1,45 @@
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+@RequestMapping("/reservations")
+public class ReservationController {
```
현재 ReservationController의 메서드는 전부 rest api이고, 앞으로도 그럴 것 같아요.
그러면 `@RestController`라고 명시해주는 것이 좋을까요?

우선 `@RestController`를 사용하지 않은 이유는
`@RestController`의 유무와 상관 없이 상태 지정을 위해 ResponseEntity를 반환할 것이고,
ResponseEntity를 반환한다면 `@RestController`를 사용하지 않아도 제 의도대로 잘 동작하기 때문입니다.

하지만 `@RestController`를 달아주면 이 Controller가 rest api만 다룬다는 것이 명확해지기 때문에 가독성이 좋아질 것 같습니다.

dusdn1702 replied:
말씀 주신 것처럼 현재 상황에서 `@RestController`를 사용할 이유는 없어보여요. 추후에 요구사항에 맞게 적용하는 게 좋을 것 같아요~

kjyyjk committed:
refactor: 예약 Request Dto에 변환 로직 추가

kjyyjk committed:
refactor: Reservation#isIdEquals() 생성

dusdn1702 commented on src/main/java/roomescape/AdminController.java:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+
+@Controller
```
AdminController와 ReservationController의 분리 좋네요 👍

dusdn1702 commented on src/main/java/roomescape/Reservation.java:
```diff
@@ -0,0 +1,41 @@
+
+public class Reservation {
+    private final long id;
+    private String name;
```
name과 dateTime도 불변값으로 보이는데, final 선언을 안하신 이유가 있나요?

kjyyjk replied:
예약자 이름과 예약 시간은 추후 변경될 수 있는 여지가 많다고 판단해서 가변값으로 두었습니다!
(예약 정보 수정 기능과 같은)

dusdn1702 replied:
좋습니다 ㅎㅎ 다만, 현재 요구사항에 맞춰 작업하고 변경될 때 적용해나가는 방안도 같이 생각해보면 좋겠어요 :)

dusdn1702 commented on src/main/java/roomescape/Reservation.java:
```diff
@@ -0,0 +1,41 @@
+public class Reservation {
+    private final long id;
+    private String name;
+    private LocalDateTime dateTime;
```
dateTime보다는 의미있는 변수명을 사용하면 어떨까요?

kjyyjk replied:
현재 `dateTime`은 예약이 생성된 날짜인지 아니면 예약 날짜인지, 의미가 혼동될 수 있을 거 같아요!
따라서 예약 날짜 및 시간을 의미하는 `reservationDateTime`으로 수정했습니다.

같은 이유로 `name` 에서 `customerName`으로 변수명을 수정했습니다.
fb416a4463579519a2622ecae3c79b17ff438253

dusdn1702 commented on src/main/java/roomescape/Reservation.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public Reservation(long id, String name, LocalDate date, LocalTime time) {
+        this(id, name, LocalDateTime.of(date, time));
```
입력도 date, time이 분리되어 들어오고, 응답도 분리하여 보내는데 저장은 dateTime으로 하는 이유가 무엇인가요?

kjyyjk replied:
하나로 관리할 수 있는 필드를 특별한 이유 없이 두 개로 나누어 관리하는 것은 비효율적이라고 생각했습니다!

dusdn1702 replied:
오히려 이렇게 하나로 작업하고 Formatter 를 나누어 관리하는 것이 불필요한 유지보수 포인트가 늘어나는 것은 아닐까요? 고민해보면 좋겠어요!

dusdn1702 commented on src/main/java/roomescape/ReservationController.java:
```diff
@@ -0,0 +1,45 @@
+        List<ReservationResponse> responses = reservations.stream()
+                .map(ReservationResponse::from)
+                .toList();
+        return ResponseEntity.ok().body(responses);
```
전체적으로 ResponseEntity를 사용하고 있는데 사용하신 이유가 궁금해요~!

kjyyjk replied:
`ReponseEntity`는 `@ResponseBody`와 마찬가지로 반환 값을 직렬화해서 응답 본문에 담을 수 있고,
더 나아가 응답 헤더나 Http상태를 지정할 수 있다는 것을 학습했습니다.

만약 `ResponseEntity`를 사용하지 않았다면 다음과 같이 작성했을 거 같습니다.
```
@GetMapping
@ResponseBody
@ResponseStatus(HttpStatus.OK)
public List<ReservationResponse> getAll() {
    ...
}
```

하지만 위 방식은
1. 애너테이션을 하나 더 추가해야한다.
2. 헤더 정보를 설정할 수 없다.
3. 상황에 따라 다른 HttpStatus로 응답할 수 없다.

라는 세 가지 단점이 있다고 생각합니다.

따라서 ResponseEntity를 사용했습니다!

kjyyjk replied:
다른 크루의 PR을 둘러보다가 별도 지정이 없으면 200(OK)을 기본으로 응답한다는 것을 알게 됐습니다.

그렇다면 기본적으로 `@ResponseBody`을 사용하다가 특정 상황에서 ResponseEntity를 사용할 수 있을 것 같아요!
ex) OK 외 Http상태를 지정해야하는 경우 ResponseEntity사용

dusdn1702 replied:
현재의 요구사항에서 Http 상태를 추가하는 상황은 없어서 여쭈어보았어요.
이유를 안다면 사용에 문제될 것 없다고 생각해요 👍

dusdn1702 commented on src/main/java/roomescape/ReservationController.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<ReservationResponse> create(@RequestBody ReservationRequest request) {
```
`@RequestBody` 도 적용하신 이유가 궁금해요!

kjyyjk replied:
`@RequestBody`를 사용하면 요청 본문에 담긴 Json데이터를 역직렬화해서 ReservationRequest로 변환한다고 학습했습니다.

만약 @RequestBody를 사용하지 않는다면,
요청 본문을 읽어온 뒤 `ObjectMapper`를 활용해 직접 ReservationRequest로 변환해줘야 하는 것으로 알고 있습니다.

내부적으로는 같은 방식이라고 생각했기 때문에 더 간편한 방법인 `@RequestBody`를 사용했습니다.

dusdn1702 replied:
좋네용 ㅎㅎ

dusdn1702 commented on src/main/java/roomescape/ReservationController.java:
```diff
@@ -0,0 +1,45 @@
+        if (isRemoved) {
+            return ResponseEntity.ok().build();
+        }
+        return ResponseEntity.notFound().build();
```
notFound로 응답하고자 하신 이유가 무엇인가요??

kjyyjk replied:
요청 시 경로 변수로 지정한 리소스가 존재하지 않기 때문에, notFound(404)로 응답하는 것이 적절하다고 생각했습니다.

실제로 HTTP Status에서 404가 무엇을 의미하는 지를 찾아보았었고,
`요청 URL이 존재하지 않는 경우` 혹은 `대상 리소스가 존재하지 않는 경우`를 의미함을 확인했습니다.

dusdn1702 replied:
추후에 생각할 부분이긴 하지만, 이러한 경우에 404 응답이 적절할 지에 대해서도 같이 생각해보면 좋겠어요.
클라이언트 입장에서 404로 응답을 받으면 어떻게 처리하게 될까요? 
지금은 어려운 부분일 수 있으니, 생각만 해두면 좋겠어요. 👍

dusdn1702 commented on src/main/java/roomescape/ReservationResponse.java:
```diff
@@ -0,0 +1,17 @@
+        long id,
+        String name,
+        LocalDate date,
+        @JsonFormat(pattern = "HH:mm") LocalTime time
```
`@JsonFormat`을 사용하신 이유가 무엇인가요?

kjyyjk replied:
요구 사항에서는 `시각:분`까지만 출력하기를 요구하지만, 실제로는 `시각:분:초`로 출력 되는 것을 확인했습니다.
이에 대해 문자열로 다루기보다는 직렬화의 형식을 따로 지정해주면 되겠다고 생각했습니다.

이후 `@JsonFormat`을 사용하면 직렬화되는 문자열의 형식을 지정해줄 수 있다는 것을 알게 됐습니다.

이 과정에서 `@DateTimeFormat`도 사용해보았었는데 의도한 대로 동작하지 않음을 확인했습니다.
관련해서 찾아본 결과 `@DateTimeFormat`은 스프링 진영의 기술이기 때문에,
Json 직렬화/역직렬화를 담당하는 Jackson 라이브러리에서는 이를 인식하지 못한다는 것을 학습했습니다.

따라서 이번 문제 상황에서는 `@JsonFormat`을 사용했습니다.

dusdn1702 commented on src/main/resources/static/index.html:
```diff
@@ -0,0 +1 @@
+Welcome, Page!
```
👍

dusdn1702 commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -1,20 +1,113 @@
+    }
+
+    @Test
+    @DisplayName("/reservations DELETE 요청에 존재하지 않는 자원이면 404로 응답한다")
```
예외 상황 테스트까지 좋네요.

dusdn1702 commented on src/main/java/roomescape/ReservationController.java:
```diff
@@ -0,0 +1,45 @@
+
+@Controller
+@RequestMapping("/reservations")
+public class ReservationController {
```
전체적으로 로직을 가진 Controller와 Reservation에 대한 단위테스트도 추가되면 좋을 것 같아요.

kjyyjk replied:
단위 테스트 추가했습니다!
957655f5d81169a83dba453248aa23044ca5964b

kjyyjk replied:
추가로 `ReservationController`에 대한 단위 테스트를 작성하다 보니,
하나의 테스트 안에서 Http 응답을 검증하고 실제 도메인 로직이 잘 수행 됐는지를 검증을 해주어야 했습니다.

그리고 ReservationController가 두 가지 책임을 지니고 있다는 것을 느꼈지만 아직

dusdn1702 reviewed (changes requested):
안녕하세요 레오~
레벨2 첫 미션 구현하시느라 고생 많으셨습니다. 요구사항에 맞게 잘 구현해주셨어요.
몇가지 궁금증과 코멘트 남겨두었습니다. 확인 부탁드릴게요!

kjyyjk committed:
style: Reservation의 변수명 수정

- 기존에는 name과 dateTime이었으나 무슨 이름과 시간을 의미하는 것인지 모호하다.
customerName과 reservationDateTime으로 수정해서 명확하게 의미를 전달한다.

kjyyjk committed:
test: ReservationController, Reservation 단위 테스트

dusdn1702 commented on src/test/java/roomescape/unit/ReservationTest.java:
```diff
@@ -0,0 +1,23 @@
+import org.junit.jupiter.api.Test;
+import roomescape.Reservation;
+
+class ReservationTest {
```
단위테스트 추가 좋네요 💯

dusdn1702 reviewed (approved):
안녕하세요 레오
피드백 반영해주신 부분들 확인했습니다.
고생 많으셨어요. 남겨둔 간단한 피드백들은 다음 단계에서 같이 학습해나가면 좋을 것 같아요.
이번 단계는 마무리 하도록 하겠습니다. 추가로 궁금하신 점 있으시면 편하게 말씀 주세요 😄
